README.md
# Python Cielo API 3.0

Integração **simplificada** do python com a [API 3.0 da Cielo](http://developercielo.github.io/Webservice-3.0/).
A integração permite compra com cartão de crédito e personalização de captura
automática, número de parcelas, etc.

## Instalação

```
pip install pythoncielo3
```

## Criando e capturando uma compra com cartão de crédito
```python
    from pythoncielo3 import Transaction

    # Os valores abaixo são informados pela Cielo. As chaves do ambiente de teste
    # podem ser obtidas em: https://cadastrosandbox.cieloecommerce.cielo.com.br
    merchant_id = 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx'
    merchant_key = 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX'

    # Para fazer transação fora do ambiente de testes remova sandbox=True da
    # chamada abaixo.
    trans = Transaction(merchant_id, merchant_key, sandbox=True)

    trans.merchant_order_id = '0001'
    trans.amount = 10000
    trans.set_customer(name='Pedro Costa')
    trans.set_credit_card(card_number='0000000000000001', holder='Pedro Costa',
                          security_code='123', expiration_date='12/2021',
                          brand='Visa')

    res = trans.create(capture=True)

    print(res['success'])
    print(res['raw']['Payment']['Tid'])
```

## Disclaimer

* A integração não permite fazer operações de leitura na API da cielo.
* Apenas foram testadas compras com cartão de crédito.
* O código foi testado apenas com python `3.5.1`.


CHANGELOG.md
# 0.0.3

## Bug Fixes

* Fix `Header value ... must be of type str or bytes` error

# 0.0.2

## Bug Fixes

* Fix unauthorized transactions returning `success=True`


tests.py
import unittest

from pythoncielo3 import Transaction

_MERCHANT_ID = '499b1bc7-8783-4b5d-8d42-f738fb68a848'
_MERCHANT_KEY = 'UYONWSYAYBAYWLERUJUGHIWCMYYBISWHBHUBMIEB'


class TestSandbox(unittest.TestCase):
    def setUp(self):
        self.trans = Transaction(_MERCHANT_ID, _MERCHANT_KEY, sandbox=True)

    def test_create_empty_transaction(self):
        res = self.trans.create()
        errors = [e['Code'] for e in res['raw']]

        self.assertFalse(res['success'])
        self.assertIn(105, errors)  # Customer Name is required.
        self.assertIn(122, errors)  # MerchantOrderId is required.

    def test_create_basic_transaction(self):
        self.trans.merchant_order_id = '0001'
        self.trans.amount = 10000
        self.trans.set_customer(name='Pedro Costa')
        self.trans.set_credit_card(card_number='0000000000000001',
                                   holder='Pedro Costa', security_code='123',
                                   expiration_date='12/2099', brand='Visa')

        res = self.trans.create()
        card = res['raw']['Payment']['CreditCard']

        self.assertTrue(res['success'])
        self.assertEqual(res['raw']['MerchantOrderId'], '0001')
        self.assertEqual(res['raw']['Customer']['Name'], 'Pedro Costa')
        self.assertEqual(res['raw']['Payment']['Amount'], 10000)
        self.assertFalse(res['raw']['Payment']['Capture'])
        self.assertEqual(card['CardNumber'], '000000******0001')
        self.assertEqual(card['Holder'], 'Pedro Costa')
        self.assertEqual(card['ExpirationDate'], '12/2099')
        self.assertEqual(card['Brand'], 'Visa')
        self.assertFalse(card['SaveCard'])

    def test_create_copletetransaction(self):
        cli = {
            'name': 'Pedro Costa',
            'email': 'pedro@costa.com',
            'birthdate': '1991-01-02',
            'address': {
                'Street': 'Rua Teste',
                'Number': '123',
                'Complement': 'AP 123',
                'ZipCode': '12345987',
                'City': 'Rio de Janeiro',
                'State': 'RJ',
                'Country': 'BRA'
            },
            'delivery_address': {
                'Street': 'Rua Teste',
                'Number': '123',
                'Complement': 'AP 123',
                'ZipCode': '12345987',
                'City': 'Rio de Janeiro',
                'State': 'RJ',
                'Country': 'BRA'
            }
        }
        self.trans.set_customer(**cli)
        self.trans.set_credit_card(card_number='0000000000000001',
                                   holder='Pedro Costa', security_code='123',
                                   expiration_date='12/2099', brand='Visa')
        self.trans.merchant_order_id = '0001'
        self.trans.amount = 10000
        self.trans.installments = 2

        res = self.trans.create(capture=True)
        card = res['raw']['Payment']['CreditCard']

        self.assertTrue(res['success'])
        self.assertEqual(res['raw']['MerchantOrderId'], '0001')
        self.assertEqual(res['raw']['Customer']['Name'], cli['name'])
        self.assertEqual(res['raw']['Customer']['Email'], cli['email'])
        self.assertEqual(res['raw']['Customer']['Birthdate'], cli['birthdate'])
        self.assertEqual(res['raw']['Customer']['Address'], cli['address'])
        self.assertEqual(res['raw']['Customer']['DeliveryAddress'],
                         cli['delivery_address'])
        self.assertEqual(res['raw']['Payment']['Amount'], 10000)
        self.assertEqual(res['raw']['Payment']['Installments'], 2)
        self.assertTrue(res['raw']['Payment']['Capture'])
        self.assertEqual(card['CardNumber'], '000000******0001')
        self.assertEqual(card['Holder'], 'Pedro Costa')
        self.assertEqual(card['ExpirationDate'], '12/2099')
        self.assertEqual(card['Brand'], 'Visa')
        self.assertFalse(card['SaveCard'])

    def test_create_unauthorized_transaction(self):
        self.trans.merchant_order_id = '0001'
        self.trans.amount = 10000
        self.trans.set_customer(name='Pedro Costa')
        self.trans.set_credit_card(card_number='0000000000000002',
                                   holder='Pedro Costa', security_code='123',
                                   expiration_date='12/2099', brand='Visa')

        res = self.trans.create()

        self.assertFalse(res['success'])

if __name__ == '__main__':
    unittest.main()


cielo.py
# -*- coding: utf-8 -*-

# Standart libs
from random import randint
from copy import deepcopy
from json import dumps

# Related third party libs
from requests import post

_DEFAULT_TRANSACTION = {
    'MerchantOrderId': None,
    'Customer': {
        'Name': None,
    },
    'Payment': {
        'Type': 'CreditCard',
        'Amount': None,
        'Installments': 1,
        'CreditCard': {
            'CardNumber': None,
            'Holder': None,
            'ExpirationDate': None,
            'SecurityCode': None,
            'Brand': None
        }
    }
}


class Transaction(object):
    def __init__(self, merchant_id, merchant_key, sandbox=False):
        def get_write_url():
            if sandbox:
                return 'https://apisandbox.cieloeCommerce.cielo.com.br/'

            else:
                return 'https://api.cieloeCommerce.cielo.com.br/'

        self._data = deepcopy(_DEFAULT_TRANSACTION)
        self._merchant_id = merchant_id
        self._merchant_key = merchant_key
        self._api_write_url = get_write_url()
        self.request_id = randint(1, 99999999999)
        self._data['Payment']['Provider'] = 'Simulado' if sandbox else 'Cielo'

    def _to_camel_case(self, snake_case):
        return ''.join(x.title() for x in snake_case.split('_'))

    def _make_header(self):
        return {
            'MerchantId':  self._merchant_id,
            'MerchantKey':  self._merchant_key,
            'RequestId':  self.request_id,
            'Content-Type':  'application/json'
        }

    def set_customer(self, **customer):
        self._data['Customer'] = {}

        for key, val in customer.items():
            self._data['Customer'][self._to_camel_case(key)] = val

    def set_credit_card(self, **card):
        self._data['Payment']['CreditCard'] = {}

        for key, val in card.items():
            self._data['Payment']['CreditCard'][self._to_camel_case(key)] = val

    def __setattr__(self, name, value):
        editable_payment_fields = [
            'amount', 'currency', 'country', 'service_tax_amount',
            'installments', 'interest', 'authenticate'
        ]

        if name in ['merchant_order_id', 'customer', 'payment']:
            self._data[self._to_camel_case(name)] = value

        elif name in editable_payment_fields:
            self._data['Payment'][self._to_camel_case(name)] = value

        else:
            super(Transaction, self).__setattr__(name, value)

    def create(self, capture=False):
        self._data['Payment']['Capture'] = capture

        url = '%s/1/sales/' % self._api_write_url
        res = post(url, dumps(self._data), headers=self._make_header())

        return {'success':  (res.status_code == 201), 'raw':  res.json()}


pythoncielo3/__init__.py
# -*- coding: utf-8 -*-

# Standart libs
from random import randint
from copy import deepcopy
from json import dumps

# Related third party libs
from requests import post

__all__ = ['Transaction']

_DEFAULT_TRANSACTION = {
    'MerchantOrderId': None,
    'Customer': {
        'Name': None,
    },
    'Payment': {
        'Type': 'CreditCard',
        'Amount': None,
        'Installments': 1,
        'CreditCard': {
            'CardNumber': None,
            'Holder': None,
            'ExpirationDate': None,
            'SecurityCode': None,
            'Brand': None
        }
    }
}


class Transaction(object):
    def __init__(self, merchant_id, merchant_key, sandbox=False):
        def get_write_url():
            if sandbox:
                return 'https://apisandbox.cieloeCommerce.cielo.com.br/'

            else:
                return 'https://api.cieloeCommerce.cielo.com.br/'

        self._data = deepcopy(_DEFAULT_TRANSACTION)
        self._merchant_id = merchant_id
        self._merchant_key = merchant_key
        self._api_write_url = get_write_url()
        self.request_id = str(randint(1, 99999999999))
        self._data['Payment']['Provider'] = 'Simulado' if sandbox else 'Cielo'

    def _to_camel_case(self, snake_case):
        return ''.join(x.title() for x in snake_case.split('_'))

    def _make_header(self):
        return {
            'MerchantId':  self._merchant_id,
            'MerchantKey':  self._merchant_key,
            'RequestId':  self.request_id,
            'Content-Type':  'application/json'
        }

    def set_customer(self, **customer):
        self._data['Customer'] = {}

        for key, val in customer.items():
            self._data['Customer'][self._to_camel_case(key)] = val

    def set_credit_card(self, **card):
        self._data['Payment']['CreditCard'] = {}

        for key, val in card.items():
            self._data['Payment']['CreditCard'][self._to_camel_case(key)] = val

    def __setattr__(self, name, value):
        editable_payment_fields = [
            'amount', 'currency', 'country', 'service_tax_amount',
            'installments', 'interest', 'authenticate'
        ]

        if name in ['merchant_order_id', 'customer', 'payment']:
            self._data[self._to_camel_case(name)] = value

        elif name in editable_payment_fields:
            self._data['Payment'][self._to_camel_case(name)] = value

        else:
            super(Transaction, self).__setattr__(name, value)

    def create(self, capture=False):
        def was_authorized():
            if res.status_code == 201:
                return 'AuthorizationCode' in raw.get('Payment', {})

            return False

        self._data['Payment']['Capture'] = capture

        url = '%s/1/sales/' % self._api_write_url
        res = post(url, dumps(self._data), headers=self._make_header())
        raw = res.json()

        return {'success': was_authorized(), 'raw': raw}
